### PR TITLE
submissions for random sequences

### PR DESCRIPTION
**Pull request by zouroboros**

### CODE REVIEW

**Comment by zouroboros** [comment]
@s-will we have multiple problems here:
1.  this time I need a newer version of the containers package then the one that is preinstalled by travis. unfortunately it is impossible to update packages that are installed using apt with cabal. Maybe we can use a newer ghc version? 
2. The MonoShuffle program doesn't need the map package but somehow it is not compiled and so the test fails

**Comment by zouroboros** [comment]
Ok if I disable the RollingDice submission the MonoShuffle submission passes.

**Comment by zouroboros** [comment]
looking at [https://github.com/commercialhaskell/stack/blob/master/.travis.yml](https://github.com/commercialhaskell/stack/blob/master/.travis.yml) it seems to be possible to install a newer ghc version (addons: {apt: {packages: [ghc-8.0.2], sources: [hvr-ghc]}})

**Comment by s-will** [comment]
Thanks for the pointer; I will try this

**Comment by s-will** [comment]
Hey @zouroboros, this didn't work out of the box. Problably, ghc needs to be called as ghc-8.0.2 now. Could you try changing this in languages.yml?

**Comment by s-will** [comment]
Hi, sorry I guessed wrong. Looks like the apt-package installs ghc under /opt/ghc/bin/ghc-8.0.2 (which is not in the path)

**Comment by zouroboros** [comment]
Yep I the latest commit should be using the absolute paths.

**Comment by zouroboros** [comment]
According to this [https://github.com/hvr/multi-ghc-travis](https://github.com/hvr/multi-ghc-travis) you need to manually update the path

**Comment by s-will** [comment]
Ok, I can include something like
before_install:
 - export PATH=/opt/ghc/$GHCVER/bin:/opt/cabal/$CABALVER/bin:$PATH
in .travis
The names would still be ghc-8.0.2 and cabal-1.24

**Comment by s-will** [comment]
or wait, the names should be ghc and cabal again:)

**Comment by zouroboros** [comment]
Ok thanks

**Comment by zouroboros** [comment]
The implementation of the uShuffle algorithmn for k-lets may not be the most "haskelly" way (especially the wilson part), but at least it works

**Comment by zouroboros** [comment]
Also working with random numbers in Haskell was definitely more tedious than I thought. I also could find more elegant ways to work with random number generators.

**Comment by s-will** [comment]
Hey, it is really cool that you implemented even the uShuffle algo in Haskell :). This is certainly challenging. I am already curious to look at your solution---currently, I just must postpone this a little.

File: A4/zouroboros-KLetShuffle.hs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+shuffleKlets rg str k = let
+  graph = kletGraph str k
+  start = List.take (k - 1) str
+  end = lastKlet str (k - 1)
+  (ust, rg') = wilson rg end graph
+  in walkGraph rg' start end ust graph
+
+results :: RandomGen g => g -> String -> Int -> [String]
+results rg str k = let
+  (first, rg') = shuffleKlets rg str k
```
Comment (s-will) [inline A4/zouroboros-KLetShuffle.hs]: maybe rewrite like:
let next (_, rg') = shuffleKlets rg' str k
in map fst . tail . iterate next $ ("", rg)
to save redundancy?

File: A4/zouroboros-KLetShuffle.hs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+import System.Environment
+import Data.Tuple (swap)
+import Data.Maybe
+import Debug.Trace
+
+-- | We represent the graph using a map of a klet (string),
+-- | to a list of connected klets (list of strings).
+-- | Note that a list can contain multiple identical entries.
+-- | This allows us the represent multiple edges between the same klets.
+type Graph = Map.Map String [String]
```
Comment (s-will) [inline A4/zouroboros-KLetShuffle.hs]: ok nice; basically, this is an adjacency list representation (where you avoid using indices for nodes (by using a more complex data structure))

File: A4/zouroboros-KLetShuffle.hs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+setEdges :: String -> [String] -> Graph -> Graph
+setEdges = Map.insert
+
+removeEdge :: String -> String -> Graph -> Graph
+removeEdge a = Map.adjust (List.delete a)
+
+-- | constructs the k-let graph of the given sequence,
+-- | with k-lets oft the given size
+kletGraph :: String -> Int -> Graph
+kletGraph str n = let
```
Comment (s-will) [inline A4/zouroboros-KLetShuffle.hs]: n should rather be named k

File: A4/zouroboros-KLetShuffle.hs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  first:subsecs = chunksOf (n - 1) str
```
Comment (s-will) [inline A4/zouroboros-KLetShuffle.hs]: this looks suspiciously like it does not work for k>2

File: A4/zouroboros-KLetShuffle.hs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  (_, graph) = foldl (\(prev, g) klet ->
+    (klet, addEdge prev klet g)) (first, Map.empty) subsecs
+  in graph
+
+lastKlet :: String -> Int -> String
+lastKlet str n = last (chunksOf n str)
```
Comment (s-will) [inline A4/zouroboros-KLetShuffle.hs]: (if I understand chunksOf correctly), this does not yield the last klet in general, e.g.
last (chunksOf 3 [1,2,3,4,5,6,7,8]) = last ([[1,2,3],[4,5,6],[7,8]) = [7,8],
where you want [6,7,8]

**Comment by s-will** [comment]
I am going to merge this after Travis says ok; if you want to make updates, please open a new PR.

**Pull request merged**